File: python/mirage/core/langfuse/readdir.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

from typing import Any

from mirage.accessor.langfuse import LangfuseAccessor
from mirage.cache.index import IndexEntry
from mirage.core.hierarchy.readdir import DirListing, make_readdir
from mirage.core.hierarchy.scope import ScopeMatch
from mirage.core.langfuse.client import (fetch_dataset_items,
                                         fetch_dataset_runs, fetch_datasets,
                                         fetch_prompts, fetch_sessions,
                                         fetch_traces)
from mirage.core.langfuse.scope import TOP_LEVEL_DIRS, detect_scope
from mirage.core.render.json import jsonl_bytes


def _bounded(accessor: LangfuseAccessor, traces: list[dict[str, Any]]) -> bool:
    """Whether a trace listing may have left traces out.

    It is one page of at most ``default_trace_limit`` traces from
    ``default_from_timestamp`` on, so a full page or a set window is a
    truncated view, not the directory: cached as complete it would prove
    an older trace absent while read fetches it by id.
    """
    return (len(traces) >= accessor.config.default_trace_limit
            or bool(accessor.config.default_from_timestamp))


def _trace_entries(
        traces: list[dict[str, Any]]) -> list[tuple[str, IndexEntry]]:
    return [(f"{t.get('id', '')}.json",
             IndexEntry(
                 id=t.get("id", ""),
                 name=t.get("id", ""),
                 resource_type="langfuse/trace",
                 vfs_name=f"{t.get('id', '')}.json",
             )) for t in traces]


async def _list_traces(accessor: LangfuseAccessor,
                       match: ScopeMatch) -> DirListing:
    traces = await fetch_traces(
        accessor.api,
        limit=accessor.config.default_trace_limit,
        from_timestamp=accessor.config.default_from_timestamp,
    )
    # The list endpoint returns trace summaries while a read renders the
    # full trace with its observations, so a size here would cost one
    # fetch_trace per entry. Traces and prompts stay size-unknown until a
    # read hydrates them; the dataset .jsonl files below are sized
    # because their listing already carries every item.
    return DirListing(entries=_trace_entries(traces),
                      partial=_bounded(accessor, traces))


async def _list_sessions(accessor: LangfuseAccessor,
                         match: ScopeMatch) -> list[tuple[str, IndexEntry]]:
    sessions = await fetch_sessions(accessor.api)
    return [(s.get("id", ""),
             IndexEntry(
                 id=s.get("id", ""),
                 name=s.get("id", ""),
                 resource_type="langfuse/session",
                 vfs_name=s.get("id", ""),
             )) for s in sessions]


async def _list_session_traces(accessor: LangfuseAccessor,
                               match: ScopeMatch) -> DirListing:
    traces = await fetch_traces(
        accessor.api,
        session_id=match.slots["session_id"],
        limit=accessor.config.default_trace_limit,
        from_timestamp=accessor.config.default_from_timestamp,
    )
    return DirListing(entries=_trace_entries(traces),
                      partial=_bounded(accessor, traces))


async def _list_prompts(accessor: LangfuseAccessor,
                        match: ScopeMatch) -> list[tuple[str, IndexEntry]]:
    prompts = await fetch_prompts(accessor.api)
    seen: set[str] = set()
    entries: list[tuple[str, IndexEntry]] = []
    for p in prompts:
        prompt_name = p.get("name", "")
        if prompt_name in seen:
            continue
        seen.add(prompt_name)
        entries.append((prompt_name,
                        IndexEntry(
                            id=prompt_name,
                            name=prompt_name,
                            resource_type="langfuse/prompt",
                            vfs_name=prompt_name,
                        )))
    return entries


async def _list_prompt_versions(
        accessor: LangfuseAccessor,
        match: ScopeMatch) -> list[tuple[str, IndexEntry]]:
    prompt_name = match.slots["prompt_name"]
    prompts = await fetch_prompts(accessor.api)
    entries: list[tuple[str, IndexEntry]] = []
    for p in prompts:
        if p.get("name") != prompt_name:
            continue
        # The list endpoint returns PromptMeta, which carries every
        # version of a prompt in a `versions` array; there is no scalar
        # `version`.
        for version in sorted(p.get("versions", [])):
            filename = f"{version}.json"
            entries.append((filename,
                            IndexEntry(
                                id=f"{prompt_name}/{version}",
                                name=str(version),
                                resource_type="langfuse/prompt_version",
                                vfs_name=filename,
                            )))
    return entries


async def _list_datasets(accessor: LangfuseAccessor,
                         match: ScopeMatch) -> list[tuple[str, IndexEntry]]:
    datasets = await fetch_datasets(accessor.api)
    return [(d.get("name", ""),
             IndexEntry(
                 id=d.get("name", ""),
                 name=d.get("name", ""),
                 resource_type="langfuse/dataset",
                 vfs_name=d.get("name", ""),
             )) for d in datasets]


async def _list_dataset(accessor: LangfuseAccessor,
                        match: ScopeMatch) -> list[tuple[str, IndexEntry]]:
    dataset_name = match.slots["dataset_name"]
    # One dataset_items call per dataset directory actually entered: the
    # dataset listing carries no item payloads, so items.jsonl can only
    # be sized here, and only for datasets the caller opens.
    items = await fetch_dataset_items(accessor.api, dataset_name)
    return [
        ("items.jsonl",
         IndexEntry(
             id=f"{dataset_name}/items",
             name="items.jsonl",
             resource_type="langfuse/dataset_items",
             vfs_name="items.jsonl",
             size=len(jsonl_bytes(items)),
         )),
        ("runs",
         IndexEntry(
             id=f"{dataset_name}/runs",
             name="runs",
             resource_type="langfuse/dataset_runs_dir",
             vfs_name="runs",
         )),
    ]


async def _list_dataset_runs(
        accessor: LangfuseAccessor,
        match: ScopeMatch) -> list[tuple[str, IndexEntry]]:
    dataset_name = match.slots["dataset_name"]
    runs = await fetch_dataset_runs(accessor.api, dataset_name)
    entries: list[tuple[str, IndexEntry]] = []
    for r in runs:
        run_name = r.get("name", "")
        filename = f"{run_name}.jsonl"
        # The listing already carries the run document read() renders, so
        # each run file's exact size is free here.
        entries.append((filename,
                        IndexEntry(
                            id=run_name,
                            name=run_name,
                            resource_type="langfuse/dataset_run",
                            vfs_name=filename,
                            size=len(jsonl_bytes([r])),
                        )))
    return entries


readdir = make_readdir(
    detect_scope,
    listers={
        "traces": _list_traces,
        "sessions": _list_sessions,
        "session": _list_session_traces,
        "prompts": _list_prompts,
        "prompt": _list_prompt_versions,
        "datasets": _list_datasets,
        "dataset": _list_dataset,
        "runs": _list_dataset_runs,
    },
    static_root=tuple(TOP_LEVEL_DIRS),
)

File: typescript/packages/core/src/core/langfuse/readdir.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import type { LangfuseAccessor } from '../../accessor/langfuse.ts'
import { IndexEntry } from '../../cache/index/config.ts'
import { type DirListing, makeReaddir } from '../hierarchy/readdir.ts'
import type { ScopeMatch } from '../hierarchy/scope.ts'
import { jsonlBytes } from '../render/json.ts'
import {
  fetchDatasetItems,
  fetchDatasetRuns,
  fetchDatasets,
  fetchPrompts,
  fetchSessions,
  fetchTraces,
} from './client.ts'
import { TOP_LEVEL_DIRS, detectScope } from './scope.ts'

// Mirrors LangfuseConfig.default_trace_limit in python.
const DEFAULT_TRACE_LIMIT = 100

function pickString(record: Record<string, unknown>, key: string): string {
  const value = record[key]
  return typeof value === 'string' ? value : ''
}

function promptVersions(record: Record<string, unknown>): string[] {
  const value = record.versions
  if (!Array.isArray(value)) return []
  const numbers: number[] = []
  for (const entry of value) {
    const parsed = typeof entry === 'number' ? entry : Number(entry)
    if (Number.isFinite(parsed)) numbers.push(parsed)
  }
  return numbers.sort((a, b) => a - b).map(String)
}

/**
 * Whether a trace listing may have left traces out.
 *
 * It is one page of at most `defaultTraceLimit` traces from
 * `defaultFromTimestamp` on, so a full page or a set window is a truncated
 * view, not the directory: cached as complete it would prove an older
 * trace absent while read fetches it by id. Mirrors python's `_bounded`.
 */
function bounded(accessor: LangfuseAccessor, traces: readonly Record<string, unknown>[]): boolean {
  const limit = accessor.config.defaultTraceLimit ?? DEFAULT_TRACE_LIMIT
  const from = accessor.config.defaultFromTimestamp
  return traces.length >= limit || (from !== undefined && from !== '')
}

function traceEntries(traces: readonly Record<string, unknown>[]): [string, IndexEntry][] {
  return traces.map((t): [string, IndexEntry] => {
    const traceId = pickString(t, 'id')
    const filename = `${traceId}.json`
    return [
      filename,
      new IndexEntry({
        id: traceId,
        name: traceId,
        resourceType: 'langfuse/trace',
        vfsName: filename,
      }),
    ]
  })
}

async function listTraces(accessor: LangfuseAccessor, _match: ScopeMatch): Promise<DirListing> {
  const limit = accessor.config.defaultTraceLimit ?? DEFAULT_TRACE_LIMIT
  // No implicit time window: an unset defaultFromTimestamp lists whatever the
  // project holds, up to defaultTraceLimit. A rolling default would hide
  // traces that read() happily serves, and python applies no window either.
  const opts: { limit: number; fromTimestamp?: string } = { limit }
  const from = accessor.config.defaultFromTimestamp
  if (from !== undefined && from !== '') opts.fromTimestamp = from
  const traces = await fetchTraces(accessor.transport, opts)
  // The list endpoint returns trace summaries while a read renders the full
  // trace with its observations, so a size here would cost one fetchTrace per
  // entry. Traces and prompts stay size-unknown until a read hydrates them;
  // the dataset .jsonl files are sized because their listing already carries
  // every item.
  return { entries: traceEntries(traces), seeds: {}, partial: bounded(accessor, traces) }
}

async function listSessions(
  accessor: LangfuseAccessor,
  _match: ScopeMatch,
): Promise<[string, IndexEntry][]> {
  const sessions = await fetchSessions(accessor.transport)
  return sessions.map((s): [string, IndexEntry] => {
    const sessionId = pickString(s, 'id')
    return [
      sessionId,
      new IndexEntry({
        id: sessionId,
        name: sessionId,
        resourceType: 'langfuse/session',
        vfsName: sessionId,
      }),
    ]
  })
}

async function listSessionTraces(
  accessor: LangfuseAccessor,
  match: ScopeMatch,
): Promise<DirListing> {
  const limit = accessor.config.defaultTraceLimit ?? DEFAULT_TRACE_LIMIT
  const opts: { sessionId: string; limit: number; fromTimestamp?: string } = {
    sessionId: match.slots.session_id ?? '',
    limit,
  }
  const from = accessor.config.defaultFromTimestamp
  if (from !== undefined && from !== '') opts.fromTimestamp = from
  const traces = await fetchTraces(accessor.transport, opts)
  return { entries: traceEntries(traces), seeds: {}, partial: bounded(accessor, traces) }
}

async function listPrompts(
  accessor: LangfuseAccessor,
  _match: ScopeMatch,
): Promise<[string, IndexEntry][]> {
  const prompts = await fetchPrompts(accessor.transport)
  const seen = new Set<string>()
  const entries: [string, IndexEntry][] = []
  for (const p of prompts) {
    const promptName = pickString(p, 'name')
    if (seen.has(promptName)) continue
    seen.add(promptName)
    entries.push([
      promptName,
      new IndexEntry({
        id: promptName,
        name: promptName,
        resourceType: 'langfuse/prompt',
        vfsName: promptName,
      }),
    ])
  }
  return entries
}

async function listPromptVersions(
  accessor: LangfuseAccessor,
  match: ScopeMatch,
): Promise<[string, IndexEntry][]> {
  const promptName = match.slots.prompt_name ?? ''
  const prompts = await fetchPrompts(accessor.transport)
  const entries: [string, IndexEntry][] = []
  for (const p of prompts) {
    if (pickString(p, 'name') !== promptName) continue
    // The list endpoint returns PromptMeta, which carries every version of a
    // prompt in a `versions` array; there is no scalar `version`.
    for (const version of promptVersions(p)) {
      const filename = `${version}.json`
      entries.push([
        filename,
        new IndexEntry({
          id: `${promptName}/${version}`,
          name: version,
          resourceType: 'langfuse/prompt_version',
          vfsName: filename,
        }),
      ])
    }
  }
  return entries
}

async function listDatasets(
  accessor: LangfuseAccessor,
  _match: ScopeMatch,
): Promise<[string, IndexEntry][]> {
  const datasets = await fetchDatasets(accessor.transport)
  return datasets.map((d): [string, IndexEntry] => {
    const datasetName = pickString(d, 'name')
    return [
      datasetName,
      new IndexEntry({
        id: datasetName,
        name: datasetName,
        resourceType: 'langfuse/dataset',
        vfsName: datasetName,
      }),
    ]
  })
}

async function listDataset(
  accessor: LangfuseAccessor,
  match: ScopeMatch,
): Promise<[string, IndexEntry][]> {
  const datasetName = match.slots.dataset_name ?? ''
  // One dataset_items call per dataset directory actually entered: the
  // dataset listing carries no item payloads, so items.jsonl can only be
  // sized here, and only for datasets the caller opens.
  const items = await fetchDatasetItems(accessor.transport, datasetName)
  return [
    [
      'items.jsonl',
      new IndexEntry({
        id: `${datasetName}/items`,
        name: 'items.jsonl',
        resourceType: 'langfuse/dataset_items',
        vfsName: 'items.jsonl',
        size: jsonlBytes(items).byteLength,
      }),
    ],
    [
      'runs',
      new IndexEntry({
        id: `${datasetName}/runs`,
        name: 'runs',
        resourceType: 'langfuse/dataset_runs_dir',
        vfsName: 'runs',
      }),
    ],
  ]
}

async function listDatasetRuns(
  accessor: LangfuseAccessor,
  match: ScopeMatch,
): Promise<[string, IndexEntry][]> {
  const datasetName = match.slots.dataset_name ?? ''
  const runs = await fetchDatasetRuns(accessor.transport, datasetName)
  return runs.map((r): [string, IndexEntry] => {
    const runName = pickString(r, 'name')
    const filename = `${runName}.jsonl`
    // The listing already carries the run document read() renders, so each
    // run file's exact size is free here.
    return [
      filename,
      new IndexEntry({
        id: runName,
        name: runName,
        resourceType: 'langfuse/dataset_run',
        vfsName: filename,
        size: jsonlBytes([r]).byteLength,
      }),
    ]
  })
}

export const readdir = makeReaddir<LangfuseAccessor>(detectScope, {
  listers: {
    traces: listTraces,
    sessions: listSessions,
    session: listSessionTraces,
    prompts: listPrompts,
    prompt: listPromptVersions,
    datasets: listDatasets,
    dataset: listDataset,
    runs: listDatasetRuns,
  },
  staticRoot: TOP_LEVEL_DIRS,
})
